core/metacling/src/TClingCallFunc.cxx

```diff
@@ -383,9 +383,17 @@ void TClingCallFunc::make_narg_call(const std::string &return_type, const unsign
          callbuf << "*(" << type_name.c_str() << "**)args["
                  << i << "]";
       } else {
-         // pointer falls back to non-pointer case; the argument preserves
-         // the "pointerness" (i.e. doesn't reference the value).
-         callbuf << "*(" << type_name.c_str() << "*)args[" << i << "]";
+         // By-value construction: Figure out if the type can be copy-constructed. This is a bit tricky in the presence
+         // of templates and templated constructors. The right way would be to ask Clang directly with the
+         // __is_constructible type trait, which triggers all of the instantiation machinery internally. However, we
+         // cannot use it in preprocessor checks, so instead use a cast with a std::conditional in the generated code
+         // and put together the right types manually.
+         callbuf << "static_cast<std::conditional<std::is_copy_constructible<" << type_name << ">::value,";
+         // If the type is copy constructible, then keep the type and let Cling copy-construct the argument:
+         callbuf << "const " << type_name << "&,";
+         // Otherwise cast to the xvalue type (mimicking std::move) to move-construct the argument:
+         callbuf << type_name << "&&";
+         callbuf << ">::type>(*(" << type_name.c_str() << "*)args[" << i << "])";
```

Comment:
Because there doesn't seem to be a way to access the information from `CXXRecordDecl`, see https://github.com/root-project/root/pull/14426. In the end, we would require to create our own type trait and then we might as well do that leveraging Cling. If you find a way to query the AST and still pass the tests I added, I'm all ears.

Reply:
Maybe we should look into how `__is_trivially_copyable` and its friends are implemented in clang?

Reply:
Yes, and it goes via the same machinery. Did you read the comment I put a couple of lines above?

Reply:
Ok... Why do you need `__is_constructible` in the preprocessor?

Reply:
To use it in `#if __is_constructible`, how else would you use it?

Can you please concretely describe which other approach you want to see tried? I thought about this for quite some while, and I see no other way than using some sort of SFINAE, which is what I implemented here.

Reply:
My worry is that this will slow down the compile time performance of almost all callfuncs. The sfinae requires template instantiations which also will cause memory growth. I think I am proposing to look into the AST operations that enable the `__is_constructible` builtin (iirc, somewhere [here](https://github.com/root-project/root/blob/ab46cb82f9d9446f7c4414a74cecd62a1b4bf371/interpreter/llvm-project/clang/lib/Sema/SemaExprCXX.cpp#L5492)), then maybe copy them over and avoid altogether the string operations. Am I missing something?

Reply:
So in the end it turns out that `__is_constructible` doesn't get us the full way, also because for example `std::vector<std::unique_ptr<T>>` always has a copy-constructor that fails to instantiate if `T` is only moveable. So for now, go back to Vincenzo's approach of iterating all constructors, find the copy constructor, and check `isDeleted`.